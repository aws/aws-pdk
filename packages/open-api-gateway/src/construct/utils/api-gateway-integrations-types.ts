// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { AuthorizationType, CorsOptions } from "aws-cdk-lib/aws-apigateway";
import { IFunction } from "aws-cdk-lib/aws-lambda";

/**
 * Defines an integration for an individual API operation
 */
export interface OpenApiIntegration {
  /**
   * The lambda function to service the api operation
   */
  readonly function: IFunction;
}

/**
 * A mapping of operation id to the integration for that operation
 */
export type OpenApiIntegrations = {
  readonly [operationId: string]: OpenApiIntegration;
};

/**
 * Structure to contain an API operation's method and path
 */
export interface MethodAndPath {
  /**
   * The path of this operation in the api
   */
  readonly path: string;
  /**
   * The http method of this operation
   */
  readonly method: string;
}

/**
 * Type for the generated Operation Lookup structure, providing details about the method and path of each API operation
 */
export type OperationLookup = {
  readonly [operationId: string]: MethodAndPath;
};

/**
 * Auth types supported by the construct
 * TODO: Support cognito auth!
 */
export type SupportedAuthTypes = AuthorizationType.IAM | AuthorizationType.NONE;

/**
 * Options required alongside an Open API specification to create API Gateway resources
 */
export interface OpenApiOptions {
  /**
   * A mapping of API operation to its integration
   */
  readonly integrations: OpenApiIntegrations;
  /**
   * Details about each operation
   */
  readonly operationLookup: OperationLookup;
  /**
   * The authorization type to use for the API
   */
  readonly authType?: SupportedAuthTypes;
  /**
   * Cross Origin Resource Sharing options for the API
   */
  readonly corsOptions?: CorsOptions;
}
