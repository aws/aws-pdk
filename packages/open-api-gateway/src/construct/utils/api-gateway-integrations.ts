/*********************************************************************************************************************
 Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

 Licensed under the Apache License, Version 2.0 (the "License").
 You may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 ******************************************************************************************************************** */

import { Stack } from "aws-cdk-lib";
import {
  AuthorizationType,
  Cors,
  CorsOptions,
  PassthroughBehavior,
  ResponseType,
} from "aws-cdk-lib/aws-apigateway";
import { Construct } from "constructs";
import { OpenAPIV3 } from "openapi-types";
import {
  MethodAndPath,
  OpenApiIntegrations,
  OpenApiOptions,
} from "./api-gateway-integrations-types";

const concatMethodAndPath = ({ method, path }: MethodAndPath) =>
  `${method.toLowerCase()}||${path.toLowerCase()}`;

/**
 * Adds API Gateway integrations and auth to the given operation
 */
const applyMethodIntegration = (
  scope: Construct,
  path: string,
  method: string,
  { integrations, authType, corsOptions }: OpenApiOptions,
  operation: OpenAPIV3.OperationObject,
  getOperationName: (methodAndPath: MethodAndPath) => string
): OpenAPIV3.OperationObject | undefined => {
  const operationName = getOperationName({ method, path });
  if (!(operationName in integrations)) {
    throw new Error(
      `Missing required integration for operation ${operationName} (${method} ${path})`
    );
  }

  const {
    function: { functionArn },
  } = integrations[operationName as keyof OpenApiIntegrations];

  // Generate the lambda invocation arn as the uri for the API Gateway integration
  const stack = Stack.of(scope);
  const uri = `arn:${stack.partition}:apigateway:${stack.region}:lambda:path/2015-03-31/functions/${functionArn}/invocations`;

  return {
    ...operation,
    responses: Object.fromEntries(
      Object.entries(operation.responses).map(([statusCode, response]) => [
        statusCode,
        {
          ...response,
          headers: {
            ...(corsOptions ? getCorsHeaderDefinitions() : {}),
            // TODO: Consider following response header references
            ...(response as OpenAPIV3.ResponseObject).headers,
          },
        },
      ])
    ),
    // https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-swagger-extensions-integration.html
    "x-amazon-apigateway-integration": {
      type: "AWS_PROXY",
      httpMethod: "POST",
      uri,
      passthroughBehavior: PassthroughBehavior.WHEN_NO_MATCH,
    },
    "x-amazon-apigateway-auth": {
      type: authType || AuthorizationType.NONE,
    },
  } as any;
};

const getCorsHeaderDefinitions = (): {
  [name: string]: OpenAPIV3.HeaderObject;
} => ({
  "Access-Control-Allow-Origin": {
    schema: { type: "string" },
  },
  "Access-Control-Allow-Methods": {
    schema: { type: "string" },
  },
  "Access-Control-Allow-Headers": {
    schema: { type: "string" },
  },
});

const generateCorsResponseHeaders = (
  corsOptions: CorsOptions
): { [key: string]: string } => ({
  "Access-Control-Allow-Headers": `'${(
    corsOptions.allowHeaders || Cors.DEFAULT_HEADERS
  ).join(",")}'`,
  "Access-Control-Allow-Methods": `'${(
    corsOptions.allowMethods || Cors.ALL_METHODS
  ).join(",")}'`,
  "Access-Control-Allow-Origin": `'${corsOptions.allowOrigins.join(",")}'`,
});

const generateCorsResponseParameters = (
  corsOptions: CorsOptions,
  prefix: string = "method.response.header"
): { [key: string]: string } =>
  Object.fromEntries(
    Object.entries(generateCorsResponseHeaders(corsOptions)).map(
      ([header, value]) => [`${prefix}.${header}`, value]
    )
  );

/**
 * Generates an "options" method with no auth to respond with the appropriate headers if cors is enabled
 */
const generateCorsOptionsMethod = (
  pathItem: OpenAPIV3.PathItemObject,
  { corsOptions }: OpenApiOptions
): OpenAPIV3.PathItemObject => {
  // Do not generate if already manually defined, or cors not enabled
  if (OpenAPIV3.HttpMethods.OPTIONS in pathItem || !corsOptions) {
    return {};
  }

  const statusCode = corsOptions.statusCode || 204;

  return {
    [OpenAPIV3.HttpMethods.OPTIONS]: {
      summary: "CORS Support",
      description: "Enable CORS by returning the correct headers",
      responses: {
        [`${statusCode}`]: {
          description: "Default response for CORS method",
          headers: getCorsHeaderDefinitions(),
          content: {},
        },
      },
      // @ts-ignore Ignore apigateway extensions which are not part of default openapi spec type
      "x-amazon-apigateway-integration": {
        type: "mock",
        requestTemplates: {
          "application/json": `{"statusCode": ${statusCode}}`,
        },
        responses: {
          default: {
            statusCode: `${statusCode}`,
            responseParameters: generateCorsResponseParameters(corsOptions),
            responseTemplates: {
              "application/json": "{}",
            },
          },
        },
      },
    },
  };
};

/**
 * Prepares a given api path by adding integrations, configuring auth
 */
const preparePathSpec = (
  scope: Construct,
  path: string,
  pathItem: OpenAPIV3.PathItemObject,
  options: OpenApiOptions,
  getOperationName: (methodAndPath: MethodAndPath) => string
): OpenAPIV3.PathItemObject => {
  return {
    ...pathItem,
    ...Object.fromEntries(
      Object.values(OpenAPIV3.HttpMethods)
        .filter((method) => pathItem[method])
        .map((method) => [
          method,
          applyMethodIntegration(
            scope,
            path,
            method,
            options,
            pathItem[method]!,
            getOperationName
          ),
        ])
    ),
    // Generate an 'options' method required for CORS preflight requests if cors is enabled
    ...generateCorsOptionsMethod(pathItem, options),
  };
};

/**
 * Prepares the api spec for deployment by adding integrations, configuring auth, etc
 */
export const prepareApiSpec = (
  scope: Construct,
  spec: OpenAPIV3.Document,
  options: OpenApiOptions
): OpenAPIV3.Document => {
  // Reverse lookup for the operation name given a method and path
  const operationNameByPath = Object.fromEntries(
    Object.entries<MethodAndPath>(options.operationLookup).map(
      ([operationName, methodAndPath]) => [
        concatMethodAndPath(methodAndPath),
        operationName,
      ]
    )
  );
  const getOperationName = (methodAndPath: MethodAndPath) =>
    operationNameByPath[concatMethodAndPath(methodAndPath)];

  return {
    ...spec,
    // https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-swagger-extensions-request-validators.html
    "x-amazon-apigateway-request-validators": {
      all: {
        validateRequestBody: true,
        validateRequestParameters: true,
      },
    },
    "x-amazon-apigateway-request-validator": "all",
    // https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-swagger-extensions-gateway-responses.html
    "x-amazon-apigateway-gateway-responses": {
      [ResponseType.BAD_REQUEST_BODY.responseType]: {
        statusCode: 400,
        responseTemplates: {
          "application/json":
            '{"message": "$context.error.validationErrorString"}',
        },
        ...(options.corsOptions
          ? {
              responseParameters: generateCorsResponseParameters(
                options.corsOptions,
                "gatewayresponse.header"
              ),
            }
          : {}),
      },
    },
    paths: {
      ...Object.fromEntries(
        Object.entries(spec.paths).map(([path, pathDetails]) => [
          path,
          preparePathSpec(scope, path, pathDetails!, options, getOperationName),
        ])
      ),
    },
  } as any;
};
