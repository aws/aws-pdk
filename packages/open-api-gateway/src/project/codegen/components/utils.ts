/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import * as fs from "fs";
import * as path from "path";
import { exec } from "projen/lib/util";
import { ClientLanguage } from "../../languages";

/**
 * Enum for generator directories for non-client generators
 */
export enum NonClientGeneratorDirectory {
  DOCS = "docs",
}

/**
 * Generator directory for openapi generation containing templates, config etc.
 */
export type GeneratorDirectory = ClientLanguage | NonClientGeneratorDirectory;

/**
 * Types of normalizers supported by openapi-generator
 * @see https://openapi-generator.tech/docs/customization/#openapi-normalizer
 */
export type OpenApiNormalizer = "KEEP_ONLY_FIRST_TAG_IN_OPERATION";

/**
 * Options for generating client code or docs using OpenAPI Generator CLI
 */
export interface GenerationOptions {
  /**
   * The OpenAPI generator to use to generate the code/docs
   */
  readonly generator: string;
  /**
   * The directory to use for OpenAPI generation
   */
  readonly generatorDirectory: GeneratorDirectory;
  /**
   * The path of the OpenAPI spec to generate the client for
   */
  readonly specPath: string;
  /**
   * The directory in which the generated code should be output
   */
  readonly outputPath: string;
  /**
   * Additional properties to pass to the generate cli
   */
  readonly additionalProperties?: {
    [key: string]: string;
  };
  /**
   * Supply the relative path from the code project root to the source code directory in which custom generated files
   * (eg. operation config) should be placed.
   */
  readonly srcDir?: string;
  /**
   * Normalizers to apply to the spec prior to generation, if any
   * @see https://openapi-generator.tech/docs/customization/#openapi-normalizer
   */
  readonly normalizers?: Partial<Record<OpenApiNormalizer, boolean>>;
}

const serializeProperties = (properties: { [key: string]: string }) =>
  Object.entries(properties)
    .map(([key, value]) => `${key}=${value}`)
    .join(",");

/**
 * Clean up any files that have already been generated by the openapi generator
 * @param generatedOutputPath output path for the generator
 */
const cleanPreviouslyGeneratedFiles = (generatedOutputPath: string) => {
  // OpenAPI generator writes a manifest called FILES which lists the files it generated.
  const openApiGeneratedFilesManifestPath = path.join(
    generatedOutputPath,
    ".openapi-generator",
    "FILES"
  );

  // If the manifest exists, delete the files it lists
  if (fs.existsSync(openApiGeneratedFilesManifestPath)) {
    const previouslyGeneratedFiles = new Set(
      fs
        .readFileSync(openApiGeneratedFilesManifestPath, { encoding: "utf-8" })
        .split("\n")
        .filter((x) => x)
    );
    previouslyGeneratedFiles.forEach((previouslyGeneratedFile) => {
      fs.unlinkSync(path.join(generatedOutputPath, previouslyGeneratedFile));
    });
  }
};

/**
 * Generate client code or docs by invoking the root generate script
 */
export const invokeOpenApiGenerator = (options: GenerationOptions) => {
  // Prior to generating, clean up any previously generated files so that we don't have stale generated code from
  // previous executions.
  cleanPreviouslyGeneratedFiles(options.outputPath);

  const srcDir = options.srcDir ?? "src";
  const additionalProperties = options.additionalProperties
    ? ` --additional-properties "${serializeProperties(
        options.additionalProperties
      )}"`
    : "";

  const normalizers = options.normalizers
    ? ` --openapi-normalizer "${serializeProperties(
        Object.fromEntries(
          Object.entries(options.normalizers).map(([k, v]) => [k, `${v}`])
        )
      )}"`
    : "";

  exec(
    `./generate --generator ${options.generator} --spec-path ${options.specPath} --output-path ${options.outputPath} --generator-dir ${options.generatorDirectory} --src-dir ${srcDir}${additionalProperties}${normalizers}`,
    {
      cwd: path.resolve(
        __dirname,
        "..",
        "..",
        "..",
        "..",
        "scripts",
        "generators"
      ),
    }
  );
};

/**
 * Options for generating documentation via a custom generator script
 */
export interface CustomDocsGenerationOptions {
  /**
   * Name of the generator script which exists in scripts/custom/docs
   */
  readonly generator: string;
  /**
   * Any arguments to pass to the script
   */
  readonly args?: string;
}

/**
 * Invoke a custom documentation generator script
 */
export const invokeCustomDocsGenerator = (
  options: CustomDocsGenerationOptions
) => {
  exec(`./${options.generator}${options.args ? ` ${options.args}` : ""}`, {
    cwd: path.resolve(
      __dirname,
      "..",
      "..",
      "..",
      "..",
      "scripts",
      "custom",
      "docs"
    ),
  });
};
