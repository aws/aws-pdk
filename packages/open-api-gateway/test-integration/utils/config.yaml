store:
  memory:
    limit: 1000

auth:
  auth-memory:
    users:
      test:
        name: test
        password: test

listen:
  - http://localhost:4873

web:
  enable: true

uplinks:
  npmjs:
    url: https://registry.npmjs.org

logs:
  - { type: stdout, format: pretty, level: warn }

packages:
  '@aws-prototyping-sdk/open-api-gateway':
    access: $all
    publish: $all
  '@*/*':
    access: $all
    publish: $all
    unpublish: $all
    proxy: npmjs
  '**':
    access: $all
    publish: $all
    unpublish: $all
    proxy: npmjs

_debug: true

max_body_size: 1gb
