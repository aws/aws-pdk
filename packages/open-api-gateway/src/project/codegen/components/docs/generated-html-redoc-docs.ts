/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import * as path from "path";
import { Component, Project } from "projen";
import { invokeCustomDocsGenerator } from "../utils";

/**
 * Configuration for the GeneratedHtmlRedocDocs component
 */
export interface GeneratedHtmlRedocDocsOptions {
  /**
   * Absolute path to the OpenAPI specification (spec.yaml)
   */
  readonly specPath: string;
}

/**
 * Generates html documentation using Redoc
 */
export class GeneratedHtmlRedocDocs extends Component {
  private options: GeneratedHtmlRedocDocsOptions;

  constructor(project: Project, options: GeneratedHtmlRedocDocsOptions) {
    super(project);
    this.options = options;
  }

  /**
   * @inheritDoc
   */
  synthesize() {
    super.synthesize();

    invokeCustomDocsGenerator({
      generator: "html-redoc",
      args: `--spec-path ${this.options.specPath} --output-path ${path.join(
        this.project.outdir,
        "html_redoc",
        "index.html"
      )}`,
    });
  }
}
