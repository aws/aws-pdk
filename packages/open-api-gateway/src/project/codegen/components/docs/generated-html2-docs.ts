/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import * as path from "path";
import { Component, Project } from "projen";
import { invokeOpenApiGenerator, NonClientGeneratorDirectory } from "../utils";

/**
 * Configuration for the GeneratedHtml2Docs component
 */
export interface GeneratedHtml2DocsOptions {
  /**
   * Absolute path to the OpenAPI specification (spec.yaml)
   */
  readonly specPath: string;
}

/**
 * Generates the html2 documentation using OpenAPI Generator
 */
export class GeneratedHtml2Docs extends Component {
  private options: GeneratedHtml2DocsOptions;

  constructor(project: Project, options: GeneratedHtml2DocsOptions) {
    super(project);
    this.options = options;
  }

  /**
   * @inheritDoc
   */
  synthesize() {
    super.synthesize();

    // Generate the html2 docs
    invokeOpenApiGenerator({
      generator: "html2",
      specPath: this.options.specPath,
      outputPath: path.join(this.project.outdir, "html2"),
      generatorDirectory: NonClientGeneratorDirectory.DOCS,
    });
  }
}
