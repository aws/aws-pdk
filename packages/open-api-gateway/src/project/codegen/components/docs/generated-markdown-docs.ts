/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import * as path from "path";
import { Component, Project } from "projen";
import { invokeOpenApiGenerator, NonClientGeneratorDirectory } from "../utils";

/**
 * Configuration for the GeneratedMarkdownDocs component
 */
export interface GeneratedMarkdownDocsOptions {
  /**
   * Absolute path to the OpenAPI specification (spec.yaml)
   */
  readonly specPath: string;
}

/**
 * Generates the markdown documentation using OpenAPI Generator
 */
export class GeneratedMarkdownDocs extends Component {
  private options: GeneratedMarkdownDocsOptions;

  constructor(project: Project, options: GeneratedMarkdownDocsOptions) {
    super(project);
    this.options = options;
  }

  /**
   * @inheritDoc
   */
  synthesize() {
    super.synthesize();

    // Generate the markdown docs
    invokeOpenApiGenerator({
      generator: "markdown",
      specPath: this.options.specPath,
      outputPath: path.join(this.project.outdir, "markdown"),
      generatorDirectory: NonClientGeneratorDirectory.DOCS,
    });
  }
}
