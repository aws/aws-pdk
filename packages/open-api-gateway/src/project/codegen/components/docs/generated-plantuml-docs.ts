/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import * as path from "path";
import { Component, Project } from "projen";
import { invokeOpenApiGenerator, NonClientGeneratorDirectory } from "../utils";

/**
 * Configuration for the GeneratedPlantUmlDocs component
 */
export interface GeneratedPlantUmlDocsOptions {
  /**
   * Absolute path to the OpenAPI specification (spec.yaml)
   */
  readonly specPath: string;
}

/**
 * Generates the plantuml documentation using OpenAPI Generator
 */
export class GeneratedPlantUmlDocs extends Component {
  private options: GeneratedPlantUmlDocsOptions;

  constructor(project: Project, options: GeneratedPlantUmlDocsOptions) {
    super(project);
    this.options = options;
  }

  /**
   * @inheritDoc
   */
  synthesize() {
    super.synthesize();

    // Generate the plantuml docs
    invokeOpenApiGenerator({
      generator: "plantuml",
      specPath: this.options.specPath,
      outputPath: path.join(this.project.outdir, "plantuml"),
      generatorDirectory: NonClientGeneratorDirectory.DOCS,
    });
  }
}
