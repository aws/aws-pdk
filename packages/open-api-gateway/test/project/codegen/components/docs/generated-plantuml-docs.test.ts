/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import { Project } from "projen";
import { GeneratedPlantUmlDocs } from "../../../../../src/project/codegen/components/docs/generated-plantuml-docs";
import { synthGeneratedCodeProject } from "../utils";

const synthDocs = (specFileName: string) => {
  const project = new Project({
    name: "parent",
  });
  return synthGeneratedCodeProject(specFileName, project, (specPath) => {
    new GeneratedPlantUmlDocs(project, {
      specPath,
    });
  });
};

describe("GeneratedPlantUmlDocs Unit Tests", () => {
  it("Single", () => {
    expect(synthDocs("single.yaml")).toMatchSnapshot();
  });

  it("Multi", () => {
    expect(synthDocs("multi.yaml")).toMatchSnapshot();
  });
});
