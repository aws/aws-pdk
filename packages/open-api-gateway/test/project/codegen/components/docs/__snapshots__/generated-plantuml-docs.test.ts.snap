// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GeneratedPlantUmlDocs Unit Tests Multi 1`] = `
Object {
  ".gitignore": "# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".
node_modules/
!/.gitattributes
!/.projen/tasks.json
!/.projen/deps.json
!/.projen/files.json
",
  ".projen/files.json": Object {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
    "files": Array [
      ".gitattributes",
      ".gitignore",
      ".projen/deps.json",
      ".projen/files.json",
      ".projen/tasks.json",
    ],
  },
  ".projen/tasks.json": Object {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
    "tasks": Object {
      "build": Object {
        "description": "Full release build",
        "name": "build",
        "steps": Array [
          Object {
            "spawn": "default",
          },
          Object {
            "spawn": "pre-compile",
          },
          Object {
            "spawn": "compile",
          },
          Object {
            "spawn": "post-compile",
          },
          Object {
            "spawn": "test",
          },
          Object {
            "spawn": "package",
          },
        ],
      },
      "compile": Object {
        "description": "Only compile",
        "name": "compile",
      },
      "default": Object {
        "description": "Synthesize project files",
        "name": "default",
      },
      "eject": Object {
        "description": "Remove projen from the project",
        "env": Object {
          "PROJEN_EJECTING": "true",
        },
        "name": "eject",
        "steps": Array [
          Object {
            "spawn": "default",
          },
        ],
      },
      "package": Object {
        "description": "Creates the distribution package",
        "name": "package",
      },
      "post-compile": Object {
        "description": "Runs after successful compilation",
        "name": "post-compile",
      },
      "pre-compile": Object {
        "description": "Prepare the project for compilation",
        "name": "pre-compile",
      },
      "test": Object {
        "description": "Run tests",
        "name": "test",
      },
    },
  },
  "plantuml/.openapi-generator-ignore": "# OpenAPI Generator Ignore
# Generated by openapi-generator https://github.com/openapitools/openapi-generator

# Use this file to prevent files from being overwritten by the generator.
# The patterns follow closely to .gitignore or .dockerignore.

# As an example, the C# client generator defines ApiClient.cs.
# You can make changes and tell OpenAPI Generator to ignore just this file by uncommenting the following line:
#ApiClient.cs

# You can match any string of characters against a directory, file or extension with a single asterisk (*):
#foo/*/qux
# The above matches foo/bar/qux and foo/baz/qux, but not foo/bar/baz/qux

# You can recursively match patterns against a directory, file or extension with a double asterisk (**):
#foo/**/qux
# This matches foo/bar/qux, foo/baz/qux, and foo/bar/baz/qux

# You can also negate patterns with an exclamation (!).
# For example, you can ignore all files in a docs folder with the file extension .md:
#docs/*.md
# Then explicitly reverse the ignore rule for a single file:
#!docs/README.md
",
  "plantuml/.openapi-generator/FILES": ".openapi-generator-ignore
schemas.plantuml
",
  "plantuml/.openapi-generator/VERSION": "6.3.0",
  "plantuml/schemas.plantuml": "@startuml

title Example API Schemas Diagram

entity ApiError {
    * errorMessage: String
}

entity TestRequest {
    myInput: BigDecimal
}

entity TestResponse {
    * messages: List<TestResponseMessagesInner>
}

entity TestResponseMessagesInner {
    message: String
    * id: Integer
}


TestResponse -- \\"0..*\\" TestResponseMessagesInner : messages

@enduml",
}
`;

exports[`GeneratedPlantUmlDocs Unit Tests Single 1`] = `
Object {
  ".gitignore": "# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".
node_modules/
!/.gitattributes
!/.projen/tasks.json
!/.projen/deps.json
!/.projen/files.json
",
  ".projen/files.json": Object {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
    "files": Array [
      ".gitattributes",
      ".gitignore",
      ".projen/deps.json",
      ".projen/files.json",
      ".projen/tasks.json",
    ],
  },
  ".projen/tasks.json": Object {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
    "tasks": Object {
      "build": Object {
        "description": "Full release build",
        "name": "build",
        "steps": Array [
          Object {
            "spawn": "default",
          },
          Object {
            "spawn": "pre-compile",
          },
          Object {
            "spawn": "compile",
          },
          Object {
            "spawn": "post-compile",
          },
          Object {
            "spawn": "test",
          },
          Object {
            "spawn": "package",
          },
        ],
      },
      "compile": Object {
        "description": "Only compile",
        "name": "compile",
      },
      "default": Object {
        "description": "Synthesize project files",
        "name": "default",
      },
      "eject": Object {
        "description": "Remove projen from the project",
        "env": Object {
          "PROJEN_EJECTING": "true",
        },
        "name": "eject",
        "steps": Array [
          Object {
            "spawn": "default",
          },
        ],
      },
      "package": Object {
        "description": "Creates the distribution package",
        "name": "package",
      },
      "post-compile": Object {
        "description": "Runs after successful compilation",
        "name": "post-compile",
      },
      "pre-compile": Object {
        "description": "Prepare the project for compilation",
        "name": "pre-compile",
      },
      "test": Object {
        "description": "Run tests",
        "name": "test",
      },
    },
  },
  "plantuml/.openapi-generator-ignore": "# OpenAPI Generator Ignore
# Generated by openapi-generator https://github.com/openapitools/openapi-generator

# Use this file to prevent files from being overwritten by the generator.
# The patterns follow closely to .gitignore or .dockerignore.

# As an example, the C# client generator defines ApiClient.cs.
# You can make changes and tell OpenAPI Generator to ignore just this file by uncommenting the following line:
#ApiClient.cs

# You can match any string of characters against a directory, file or extension with a single asterisk (*):
#foo/*/qux
# The above matches foo/bar/qux and foo/baz/qux, but not foo/bar/baz/qux

# You can recursively match patterns against a directory, file or extension with a double asterisk (**):
#foo/**/qux
# This matches foo/bar/qux, foo/baz/qux, and foo/bar/baz/qux

# You can also negate patterns with an exclamation (!).
# For example, you can ignore all files in a docs folder with the file extension .md:
#docs/*.md
# Then explicitly reverse the ignore rule for a single file:
#!docs/README.md
",
  "plantuml/.openapi-generator/FILES": ".openapi-generator-ignore
schemas.plantuml
",
  "plantuml/.openapi-generator/VERSION": "6.3.0",
  "plantuml/schemas.plantuml": "@startuml

title Example API Schemas Diagram

entity ApiError {
    * errorMessage: String
}

entity TestRequest {
    myInput: BigDecimal
}

entity TestResponse {
    * messages: List<TestResponseMessagesInner>
}

entity TestResponseMessagesInner {
    message: String
    * id: Integer
}


TestResponse -- \\"0..*\\" TestResponseMessagesInner : messages

@enduml",
}
`;
